apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -560,11 +584,84 @@ public void registerSpanListener(ActivationListener activationListener) {
         return activationListeners;
     }
 
-    void registerLifecycleListeners(List<LifecycleListener> lifecycleListeners) {
+    synchronized void start(List<LifecycleListener> lifecycleListeners) {
         this.lifecycleListeners.addAll(lifecycleListeners);
         for (LifecycleListener lifecycleListener : lifecycleListeners) {
             lifecycleListener.start(this);
         }
+        if (activeConfigOptionSet) {
+            tracerState = TracerState.RUNNING;
+            logger.info("Tracer switched to RUNNING state");
+        } else {
+            pause();
+        }
+    }
+
+    public synchronized void stressDetected() {
+        currentlyUnderStress = true;
+        pause();
+    }
+
+    public synchronized void stressRelieved() {
+        currentlyUnderStress = false;
+        if (activeConfigOptionSet) {
+            resume();
+        }
+    }
+
+    private synchronized void activeConfigChanged(boolean wasActive, boolean shouldBeActive) {
+        // if changed from true to false then:
+        //      if current state is RUNNING - pause the agent
+        //      otherwise - ignore
+        // if changed from false to true then:
+        //      if current state is RUNNING or STOPPED - no effect
+        //      if current state is PAUSED and currentlyUnderStress==false - then resume
+        if (wasActive && !shouldBeActive && tracerState == TracerState.RUNNING) {
+            pause();
+        } else if (!wasActive && shouldBeActive && tracerState == TracerState.PAUSED && !currentlyUnderStress) {
+            resume();
+        }
+        activeConfigOptionSet = shouldBeActive;
+    }
+
+    synchronized void pause() {
```

Comment:
why do we have `pause` and `resume` as package-private ? if only for testing, doing it through `stressDetected` and `stressRelieved` methods above should be enough.

Reply:
This is an (internal) tracer API, so I think package private is a good choice.
I use it widely in tests, which by itself qualifies to make it package private. 
Testing what happens when we pause the agent in the Quartz test through `stressDetected` is a much worse option in my opinion.